### PR TITLE
Display partial frames

### PR DESCRIPTION
**Pull request by egoughnour**
### The change to create_audio_output() is complete, and the video frame composition in build_output_video() is also complete.

Any thoughts about how to visualize the combination of frames?  Obviously there are a lot of potential ways to do it.

One idea I had was to slice the video frame into columns with width of `int(coefficient*frame_width/(sum of coefficients))` .
This would be fine for all but the smallest component (the smallest component with an integer number of columns assigned to it).  The last visible component would then just take all remaining pixels of total frame width--otherwise it might end up with a bad width per frame.

 Anyhow the frame that was actually used for the audio corresponding to each coefficient would be used as the source for each slice with non-zero width in the output video frame.  That should be relatively simple to slice using PIL or some slicing package or other.   All the data is computed, the frame slice widths and the video frame composition are all that need to be done.

At that point it would be functional, and even just leaving the `MAX_BASIS WIDTH`, i.e., maximum allowable number of frames in the linear combination, set to one--that ought to enough to revert to the current functionality.

I'm not sure if it would look odd to have a left-to-right composition.  It would leave the most influential parts of the image always on the left though the focus will generally have been in the middle.

Instead it might be visually appealing to rearrange the slices so that the widest slices are in the middle.  This is just based on the intuition that the center of the frame is normally the focus.

Another possible visualization would be to track the influence of each basis vector in the linear combination through time.
This would leave open the possibility of scaling the height of each basis image slice by its coefficient.
 Over a short period of time, assuming relatively gradual changes in the constituent basis vectors, it would be possible to generate a graphic equalizer effect.

### CODE REVIEW

File: stammer.py [hunk 1]
```diff
@@ -128,21 +154,22 @@ def build_output_video(frames_dir, outframes_dir, best_matches, framerate, outpu
         check=True
     )
 
-def create_output_audio(best_matches, modulator_audio, carrier_frames, modulator_frames, samples_per_frame):
+def create_output_audio(best_matches, coefficients, modulator_audio, carrier_frames, modulator_frames, samples_per_frame):
     output_audio = np.zeros(modulator_audio.shape, dtype=float)
+    def get_carrier(k,c):
+        composite_carrier = None
+        for index, element in enumerate(c):
+            if element == 0:
```
Comment (egoughnour) [inline stammer.py:162]: This ( `break if element == 0` ) means that the frame at index 0 can still be referenced.  Both the matches and coefficients are zero-padded, meaning an element (i.e., coefficient) with value of zero is at the start of the padding.
`k[index]` on the other hand could be zero as a reference to the first carrier frame.

File: stammer.py [hunk 2]
```diff
@@ -217,19 +244,22 @@ def process(carrier_path, modulator_path, output_path):
     modulator_bands = make_normalized_bands(modulator_frames, BAND_WIDTH)
 
     print("finding best matches")
-    best_matches = []
+    basis_coefficients = {}
+    best_matches_MATRIX = np.zeros((len(modulator_bands), MAX_BASIS_WIDTH), np.int32) - np.ones((len(modulator_bands), MAX_BASIS_WIDTH), np.int32)
```
Comment (egoughnour) [inline stammer.py:248]: This leaves an array filled with -1 as a sentry value when assigning matches.
These are indexes into the carrier frames so -1 is a value that should not be reached.
Yes we can use -1  as a reference relative to the end, perhaps, but a loop over the actual indexes in the array would never reach -1.

File: stammer.py [hunk 3]
```diff
@@ -257,6 +287,10 @@ def main():
         return
     try:
         TEMP_DIR.mkdir()
+    except FileExistsError:
+        shutil.rmtree(TEMP_DIR)
+        TEMP_DIR.mkdir()
+    try:
```
Comment (egoughnour) [inline stammer.py:293]: This just made debugging easier when I had to append sys.argv manually.  Then I didn't have to worry about the state of the `TEMP_DIR` directory.

File: image_tiling.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            x0, y0, w, h = tiling.get_image_placement()
+            tb = chalmers.copy()
+            tb.thumbnail((w,h))
+            output.paste(tb, (x0,y0))
+            output.save(debug_path)
+            continue
+        if c[i]:
+            x0, y0, w, h = tiling.get_image_placement()
+            tb = skinner.copy()
+            tb.thumbnail((w,h))
```
Comment (egoughnour) [inline image_tiling.py:108]: thumbnail((w,h)) doesn't change the aspect ratio.  Thus for most circumstances we ought to paste again at (x0, y0 + tb.height)

**Comment by ArdenButterfield** [comment]
I noticed you've got `from scratch_fractions.from_scratch import as_array` in stammer.py and no such file in the scratch_functions folder... Is the pull request ready to review or are you still cooking?

**Comment by egoughnour** [comment]
It worked as-is before I started simplifying it again.  There I was just removing the test code ('main'), renaming the file and then was going to move it back to the repo root directory.  It is in fraction_bits.py.  I think that might be more sensible.
```python
import fraction_bits
# ...
the_fraction = 13.0/57.0
bit_list, hot_indexes = fraction_bits.as_array(the_fraction)
```
That seems cleaner and clearer, given that it's returning the list of bits and then the indexes of the hot bits.

**Comment by ArdenButterfield** [comment]
Ah I see, just had to change the import statement. I love how the frames spiral around.

**Comment by egoughnour** [comment]
After watching the end of the _Land of Confusion_ music video (O.G., not remix) and listening to _Galvanize_ repeatedly, I came to the realization that this is probably ready for review.

**Comment by ArdenButterfield** [comment]
Sorry to take ten million years to get to the review... I love what you've got going here, and I'm excited to mess around with it more.

**Pull request merged**